### PR TITLE
feat: Associate and apply stylings to create bolded markdown

### PR DESCRIPTION
**Pull request by yoomlam**
## Ticket

Two styling-related tickets:
1. https://navalabs.atlassian.net/browse/DST-421
2. https://navalabs.atlassian.net/browse/DST-422

## Changes
- Clarified the [Workflow diagram] (https://navasage.atlassian.net/wiki/spaces/NL/pages/1305083960/Tech+Spec+for+improving+PDF+parsing#Workflow) when to associate and apply stylings
- added `associate_stylings()`
- added `apply_stylings()`
- use the new methods in `ingest_bem_pdfs._parse_pdf()` and `pdf_postprocess.add_markdown()`

## Testing

Added tests for the 2 new functions and updated `test_add_markdown()` to include a list item with styling.

### CODE REVIEW

File: app/src/ingestion/pdf_postprocess.py [hunk 1]
```diff
@@ -1,18 +1,67 @@
...
+
+
+def styling_matches_text(styling: Styling, e_text: EnrichedText) -> bool:
+    # Quick checks
+    if styling.pageno != e_text.page_number or styling.headings != e_text.headings:
+        return False
+
+    # Slower checks
+    stripped_wider_text = basic_ascii(styling.wider_text).strip()
+    stripped_e_text = basic_ascii(e_text.text).strip()
```
Comment (KevinJBoyer) [inline app/src/ingestion/pdf_postprocess.py:30]: What's the purpose of stripping to ASCII? Rendering differences between pdfminer and unstructured?

**Review by KevinJBoyer — APPROVED** [review]
LGTM, couple of questions to make sure I'm following the code and a few optional nits. Good stuff!

File: app/src/ingestion/pdf_postprocess.py [hunk 2]
```diff
@@ -1,18 +1,67 @@
...
+    return (
+        stripped_wider_text in stripped_e_text
+        and abs(len(stripped_wider_text) - len(stripped_e_text)) < 10
```
[the remark this answers is not in the document]
Comment (yoomlam) [reply app/src/ingestion/pdf_postprocess.py:33]: It's tested here: https://github.com/navapbc/labs-decision-support-tool/pull/64/files#diff-cb484d4c2a74b1561236b5b01ef2f23d96ad38d56e7068d704b5e26d4667da5dR186

There could be several paragraphs on the same page and under the same heading that have a bolded word, e.g., "CDC" (single-word paragraph) and "CDC is the center for disease control.". The styling would match both paragraphs without this check.

File: app/src/ingestion/pdf_postprocess.py [hunk 3]
```diff
@@ -1,18 +1,67 @@
...
+        for styling in e_text.stylings:
+            markdown_text = e_text.text.replace(styling.text, f"**{styling.text}**")
+            if e_text.text != markdown_text:
+                applied.append(styling)
+                e_text.text = markdown_text
+
+        if applied == e_text.stylings:
+            e_text.stylings = None
+        else:
+            e_text.stylings = [s for s in e_text.stylings if s not in applied]
```
[the remark this answers is not in the document]
Comment (yoomlam) [reply app/src/ingestion/pdf_postprocess.py:50]: Your suggestion is what I had originally but I added the `= None` line to keep the unit tests more concise and less cluttered since the unit tests compares against the original `enriched_texts`, which have `stylings = None` which is different from `stylings = []`.

File: app/src/ingestion/pdf_postprocess.py (hunk 1 above)
Comment (yoomlam) [reply app/src/ingestion/pdf_postprocess.py:30]: pdfminer extracts the bullet (`• `) from the PDF, e.g. https://github.com/navapbc/labs-decision-support-tool/blob/0323b304eee9232483f76bd8bb2067fcede9b082/app/tests/src/ingestion/test_pdf_stylings.py#L17

unstructured removes the bullet.
